File: week3/5/A_53.java
// https://leetcode.com/problems/maximum-subarray/description/
// https://walkccc.me/LeetCode/problems/0053/

// O(n)
class Solution {
  public int maxSubArray(int[] nums) {
    // dp[i] := the maximum sum subarray ending in i
    int[] dp = new int[nums.length];

    dp[0] = nums[0];
    for (int i = 1; i < nums.length; ++i)
      dp[i] = Math.max(nums[i], dp[i - 1] + nums[i]);

    return Arrays.stream(dp).max().getAsInt();
  }
}

// O(1)
class Solution {
  public int maxSubArray(int[] nums) {
    int ans = Integer.MIN_VALUE;
    int sum = 0;

    for (final int num : nums) {
      sum = Math.max(num, sum + num);
      ans = Math.max(ans, sum);
    }

    return ans;
  }
}

File: week3/5/A_53.rs
// https://leetcode.com/problems/maximum-subarray/description/

use std::cmp::max;

struct Solution{}
impl Solution {
    pub fn max_sub_array(nums: Vec<i32>) -> i32 {
        let mut answer= i32::MIN;
        let mut sum= 0;

        for num in nums  {
            sum= max(num, sum+num);
            answer= max(answer, sum);
        }

        answer
    }
}
fn main() {
    let nums= Vec::from([-2,1,-3,4,-1,2,1,-5,4]);
    let result= Solution::max_sub_array(nums);
    println!("{}", result);
}
